src/components/main-form.js

```diff
@@ -1,180 +1,192 @@
+  }
 
-    const value = parseFloat(e.target.value);
+  calculateInterestAmt(o, state) {
+    return o.pmt * (o.term * COMPOUND_FREQUENCY) + state.loanAmt;
+  }
 
-    _.set(state, key, value);
-    if (!isNaN(value)) {
-      state = this.calculate(state);
+  calculateFV(o1, o2, state) {
+    // TODO: what if o1 and o2 are equal?
+    if (o1.term > o2.term) {
+      return FV((state.investmentRate - state.inflation) / 100 / COMPOUND_FREQUENCY, o1.term * COMPOUND_FREQUENCY, o2.pmt - o1.pmt, 0);
+    } else if (o1.term < o2.term) {
+      return FV((state.investmentRate - state.inflation) / 100 / COMPOUND_FREQUENCY, (o2.term - o1.term) * COMPOUND_FREQUENCY, o1.pmt, 0);
     }
-    this.setState(state);
   }
 
-  validateData(obj) {
-    for (let key in obj)
-      if (typeof obj[key] === 'object') {
-        return this.validateData(obj[key])
-      }
-      else if (!obj[key] && obj[key] !== 0) {
-        return false;
-      }
-    return true;
+  calculateOptCost(o1, o2) {
+    return (o1.fv + o1.interestAmt) - (o2.fv + o2.interestAmt);
   }
 
-  calculate(state) {
-    const COMPOUND_FREQUENCY = 12;
+  updateInput(e) {
+    const state = Object.assign({}, this.state);
+    const key = e.target.name;
 
-    const a = state.options.a;
-    const b = state.options.b;
+    const value = parseFloat(e.target.value);
 
-    if (!a.mortgageTerm || !b.mortgageTerm) {
-      return state;
+    _.set(state, key, value);
+    if (!isNaN(value)) {
+      this.calculateAll(state);
+    } else {
+      this.setState(state);
     }
+  }
 
-    a.pmt = PMT(a.mortgageInterestRate / 100 / COMPOUND_FREQUENCY, a.mortgageTerm * COMPOUND_FREQUENCY, state.mortgageAmt);
-    a.interestAmt = a.pmt * (a.mortgageTerm * COMPOUND_FREQUENCY) + state.mortgageAmt;
+  calculateAll(state) {
+    o1 = state.options.o1;
+    o2 = state.options.o2;
 
-    b.pmt = PMT(b.mortgageInterestRate / 100 / COMPOUND_FREQUENCY, b.mortgageTerm * COMPOUND_FREQUENCY, state.mortgageAmt);
-    b.interestAmt = b.pmt * (b.mortgageTerm * COMPOUND_FREQUENCY) + state.mortgageAmt;
+    o1.pmt = this.calculatePMT(o1, state);
+    o2.pmt = this.calculatePMT(o2, state);
 
-    a.fv = FV(state.investmentRate / 100 / COMPOUND_FREQUENCY, a.mortgageTerm * COMPOUND_FREQUENCY, b.pmt - a.pmt, 0);
-    b.fv = FV(state.investmentRate / 100 / COMPOUND_FREQUENCY, (a.mortgageTerm - b.mortgageTerm) * COMPOUND_FREQUENCY, b.pmt, 0);
+    o1.interestAmt = this.calculateInterestAmt(o1, state);
+    o2.interestAmt = this.calculateInterestAmt(o2, state);
 
-    return state;
-  }
+    o1.fv = this.calculateFV(o1, o2, state);
+    o2.fv = this.calculateFV(o2, o1, state);
 
-  render() {
-    const a = this.state.options.a;
-    const b = this.state.options.b;
+    state.optCost = this.calculateOptCost(o1, o2);
 
-    let gainForA = a.fv + a.interestAmt;
-    let gainForB = b.fv + b.interestAmt;
+    this.setState(state);
```

Comment:
`calculateAll` does a `setState`, whereas the other `calculate` methods do not. This may lead to some confusion b/c I didn't expect `calculate` to update state, and instead be a pure function

Reply:
I can make it pure, no problemo